### PR TITLE
[4.3] Media Manager support for svg images + set natural{width/height}

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/39571, https://github.com/joomla/joomla-cms/issues/39538, https://github.com/joomla/joomla-cms/issues/39543 . Complimentary to https://github.com/joomla/joomla-cms/pull/39574 Previous attempt: https://github.com/joomla/joomla-cms/pull/38722

### Summary of Changes

- Fixes the wrong mime type for svg
- adds a load event to grab the naturalWidth/naturalHeight

### In pictures

Screenshot 2023-01-10 at 09 18 40
Screenshot 2023-01-10 at 09 18 51
Screenshot 2023-01-10 at 09 19 02

### Testing Instructions
- First add `svg` to the image extensions, allowed extensions and `image/svg+xml` to the allowed mime types
- Goto Caasiopeia styles and select a new logo
- Upload an svg
- Select the svg and check that the url has width/height at the end

### Expected result AFTER applying this Pull Request

Screenshot 2023-01-10 at 09 19 27

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: libraries/src/Helper/MediaHelper.php [hunk 1]
```diff
@@ -113,6 +113,10 @@ public static function getMimeType($file, $isImage = false)
             $mime = static::getMimeType($file, false);
         }
 
+        if ($mime === 'application/octet-stream' && !$isImage && strtolower(pathinfo($file)['extension']) === 'svg') {
```
Comment (dgrammatiko) [inline libraries/src/Helper/MediaHelper.php]: @laoneo @HLeithner @bembelimen are you ok with this?

Comment (laoneo) [reply libraries/src/Helper/MediaHelper.php]: Looks ok, but I'm not sure if my knowledge is good enough to detect here a security issue. You can also use PATHINFO_EXTENSION as second argument to the pathinfo call, then you get directly back the extension and not an array.

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on ed67f3586a2c69081ea50f94a502ff6d4eb1f4e4<br /><br />👌<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39586">issues.joomla.org/tracker/joomla-cms/39586</a>.</sub>

File: libraries/src/Helper/MediaHelper.php (hunk 1 above)
Comment (HLeithner) [reply libraries/src/Helper/MediaHelper.php]: Using the extension is not the best idea... 
Using the regex from https://github.com/BaseMax/detect-svg/blob/master/index.js and checking the content of the file should be better.

Comment (dgrammatiko) [reply libraries/src/Helper/MediaHelper.php]: Regex or DOMDocument?

@HLeithner in another project we use https://packagist.org/packages/league/mime-type-detection to handle the mime types more efficiently, Joomla probably should do that as well (there's also a symphony pkg)

Comment (HLeithner) [reply libraries/src/Helper/MediaHelper.php]: yeah sounds like a better idea to use a proper library.
But the library you liked looks like to be only a wrapper around finfo which we already use

Comment (dgrammatiko) [reply libraries/src/Helper/MediaHelper.php]: I guess the symphony pkg might be a better fit, especially utilising the custom sniffer
https://symfony.com/doc/current/components/mime.html#adding-a-mime-type-guesser

My point is that leveraging an existing pkg will be less code for the project to maintain, safer and requires less own questionable implementations, but this probably is out of the scope of this PR

Comment (HLeithner) [reply libraries/src/Helper/MediaHelper.php]: @SniperSister anything against the extension detection variant for now?

File: administrator/components/com_media/resources/scripts/store/mutation-types.es6.js [hunk 2]
```diff
@@ -48,5 +48,9 @@ export const RENAME_SUCCESS = 'RENAME_SUCCESS';
 // Share model
```
Comment (Quy) [inline administrator/components/com_media/resources/scripts/store/mutation-types.es6.js]:
```suggestion
// Share modal
```
I know it is unrelated to this PR, but it is not worth opening a separate PR for an obvious typo. Thanks.

**Comment by Quy** [comment]
Added to the settings, but I am getting "Unable to upload file." error. I have to investigate why. Added manually for now. 

Following are issues:

No image on the frontend.

No preview.
39586-preview

Width/Height = 0
39586-select

No dimensions.
39586-info

**Comment by dgrammatiko** [comment]
> Added to the settings, but I am getting "Unable to upload file." error. I have to investigate why. Added manually for now.

Is it a valid svg file? 
Can you share it? (eg zip it and uploaded here)
Can you clean the svg using: https://jakearchibald.github.io/svgomg/ and try again?

**Comment by Quy** [comment]
My settings are the same. Here is the cleaned logo using the service which I can now upload but the other issues remain.

[clean_logo.zip](https://github.com/joomla/joomla-cms/files/10385490/clean_logo.zip)

**Comment by Quy** [comment]
In the browser console, after selecting a SVG file for the logo.

39586-select-svg

**Comment by dgrammatiko** [comment]
Oh, I think it's the `+` on your folder that messes up the apache, did you remember that issue?

**Comment by Quy** [comment]
I do remember now, but I have the same issues under `joomla-cms-4.3-dev` without +.

**Comment by dgrammatiko** [comment]
> In the browser console, after selecting a SVG file for the logo.

Inspecting more carefully the reported URL is wrong, it shouldn't have the `/administrator` part. This comes from the Form Field Media and now I have to debug this, but to do so I need to reproduce the error...

**Comment by dgrammatiko** [comment]
@Quy can you do something, got to the page where the console error happen, open the browser inspector and paste in the console this `Joomla.getOptions('system.paths')` and report back what is the value returned. Thanks

**Comment by Quy** [comment]
> Object { root: "/joomla-cms-4.3-dev", rootFull: "http://localhost/joomla-cms-4.3-dev/", base: "/joomla-cms-4.3-dev/administrator", baseFull: "http://localhost/joomla-cms-4.3-dev/administrator/" }

**Comment by dgrammatiko** [comment]
Thanks, so the values are correct...

**Comment by Quy** [comment]
This is only happening with SVG files.

**Comment by Quy** [comment]
This could be why width/height=0 due to not finding the image when selecting a SVG file.
`images/clean_logo.svg#joomlaImage://local-images/clean_logo.svg?width=0&height=0`

**Comment by dgrammatiko** [comment]
> This could be why width/height=0 due to not finding the image when selecting a SVG file.

Yes, but the root is not actually in the Form Field Media, but probably in the PHP code of either the filesystem or the MediaHelper, I guess for some reason the followin conditional is wrong on your machine: https://github.com/joomla/joomla-cms/pull/39586/files#diff-b9cf1b73ab2438daa98290f2c650099e320527df061d94e31763982e9743dacdR116

Could you patch that line to:

```php
        if (strtolower(pathinfo($file, PATHINFO_EXTENSION)) === 'svg' && self::isValidSvg($file, false)) {
```

**Comment by Quy** [comment]
No go :(

**Comment by dgrammatiko** [comment]
Weird, can you replace the whole function with 

```php
    public static function getMimeType($file, $isImage = false)
    {
        // If we can't detect anything mime is false
        $mime = false;

        if (strtolower(pathinfo($file, PATHINFO_EXTENSION)) === 'svg' && self::isValidSvg($file, false)) {
            return 'image/svg+xml';
        }

        try {
            if ($isImage && \function_exists('exif_imagetype')) {
                $mime = image_type_to_mime_type(exif_imagetype($file));
            } elseif ($isImage && \function_exists('getimagesize')) {
                $imagesize = getimagesize($file);
                $mime      = $imagesize['mime'] ?? false;
            } elseif (\function_exists('mime_content_type')) {
                // We have mime magic.
                $mime = mime_content_type($file);
            } elseif (\function_exists('finfo_open')) {
                // We have fileinfo
                $finfo = finfo_open(FILEINFO_MIME_TYPE);
                $mime  = finfo_file($finfo, $file);
                finfo_close($finfo);
            }
        } catch (\Exception $e) {
            // If we have any kind of error here => false;
            return false;
        }

        // If we can't detect the mime try it again
        if ($mime === 'application/octet-stream' && $isImage === true) {
            $mime = static::getMimeType($file, false);
        }

        // We have a mime here
        return $mime;
    }
```

**Comment by crystalenka** [comment]
Many SVG files do not have inherent width/height set, but they are valid. Could that be why it's failing?

**Comment by dgrammatiko** [comment]
> Many SVG files do not have inherent width/height set, but they are valid. Could that be why it's failing?

Nope, the failure is a 404, file not found, probably an edge case server weirdness...

**Comment by dgrammatiko** [comment]
> Same error.

Ok, let me recap the problem here:
- the svg is uploaded through the media manager
- the image is displayed in the media manager but the sidebar doesn't have and dimensions
- selecting the image in the form field media throws a console error with a url pointing to `/administrator/images`

Do I got it correctly?

**Comment by Quy** [comment]
Yes, plus no preview in media manager and does not display on the front end.

**Comment by Quy** [comment]
Yes, that is the file.

**Comment by dgrammatiko** [comment]
Ok, so I edited the `mime.types` on my apache conf folder `/opt/homebrew/etc/httpd ` and removed the svg support like
Screenshot 2023-01-10 at 20 49 39

and got this
Screenshot 2023-01-10 at 20 51 35

and 
Screenshot 2023-01-10 at 20 52 31

Probably I could patch the media field js to skip the width and height if the values are 0, but honestly this is a server misconfiguration to me. I mean if you cannot see the images selecting them for usage is not very wise as they will not appear also on the front end...

**Comment by Quy** [comment]
I will test it on a live server to see if it is a local issue. I will report back later today or tomorrow. Thank you!

**Comment by dgrammatiko** [comment]
> I will report back later today or tomorrow

Take your time, this PR awaits the review of the security team, so nothing too urgent. Also thank you for your time here 🙌

**Comment by dgrammatiko** [comment]
@Quy what was your browser? (I should have asked that many comments earlier)

**Comment by Quy** [comment]
Windows Firefox 108.0.2, but also tested with Chrome.

**Comment by dgrammatiko** [comment]
@Quy I added fallback values for FF till I could think something more clever, now FF is working correctly but the aspect ratio (what width and size is used for) is probably wrong, so always use CSS for the sizing of SVG images...

**Comment by Quy** [comment]
Works locally/live server except for the Preview action. 👍

**Comment by Quy** [comment]
Tested on cloudaccess.host. No preview image using Chrome Version 109.0.5414.75.

29586-no-preview

**Comment by dgrammatiko** [comment]
> Tested on cloudaccess.host. No preview image using Chrome Version 109.0.5414.75.

Ok, svg images will not have a preview modal (until we find a way to get the image drawing consistently)...

**Comment by crystalenka** [comment]
Why not set an arbitrary height and width (like 60vw and 60vh) to the svg and use `object-fit: contain` to preserve the aspect ratio in cases where there is no detectable height or width?

**Comment by dgrammatiko** [comment]
> Why not set an arbitrary height and width (like 60vw and 60vh) to the svg and use object-fit: contain to preserve the aspect ratio in cases where there is no detectable height or width?

I've tried quite some tricks (also setting some hardcoded values) but the problem is that FF is not actually rendering anything. That said, I will also try your suggestion.
The other problem with the preview modal is due to some funky code (either my bad or something is unsetting the selected image, I will try to debug this the coming weekend)

If you care to dig deeper on the Firefox issue there are some links
https://bugzilla.mozilla.org/show_bug.cgi?id=1328124
https://bugzilla.mozilla.org/show_bug.cgi?id=700533
https://github.com/whatwg/html/issues/3510
Hopefully this will be resolved at some point...

As a middle ground solution I was thinking exposing 2 inputs for width/height, similar to the options we have when the image selected goes into an editor. But that's a rough idea right now, I need to see if there are problems with the existing code before proposing something...

**Comment by Quy** [comment]
The preview image is full width so it displays huge. Maybe reduce it so the title/close button will be more visible.

**Comment by crystalenka** [comment]
Since vw and vh are relative sizes it should be fairly responsive. I don't think you need to do breakpoints for this but maybe using a minmax or clamp function would help. 

like `width: clamp(300px, 1000px, 75vw)`

And then having height as 60vh-70vh ish should be fine with object fit set to contain

Forgive me if the syntax is wrong, I'm replying from my phone.

**Comment by crystalenka** [comment]
I hope that syntax worked okay 😅🙈

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on be075155dea2666c2cfb576aa4ec2e37232b7d0a<br /><br />THANK YOU!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39586">issues.joomla.org/tracker/joomla-cms/39586</a>.</sub>

**Comment by obuisard** [comment]
I have been doing some testing and I would like to know if someone got into this one:

on my tests, everything works great (yeah!) on my php 8 environments. However, uploads fail and previews don't show in php 7.4 in the media manager. I will retry with the full package (I just tried the PR's update package) and see if I get the same issues. 

Nevertheless, I wanted to mention it and see if others ran into the same problems. I don't have any errors in logs or console errors either.

**Comment by dgrammatiko** [comment]
> However, uploads fail and thumbnails don't show in php 7.4 in the media manager

Are those svg files or any image?
Also couple of things: what are the contents of mime.types of your 7.4 apache? In particular is there an uncommented line with `svg image/svg+xml`?
But this might come from the Sanitizer, which is a bug revealed but this PR and probably worth a new issue

**Comment by obuisard** [comment]
> > However, uploads fail and thumbnails don't show in php 7.4 in the media manager
> 
> Are those svg files or any image? 

Only SVG images.

> Also couple of things: what are the contents of mime.types of your 7.4 apache? In particular is there an uncommented line with `svg image/svg+xml`? 

image/svg+xml is uncommented in mime. Types.

Note that I use the same Apache for php 7 and 8 tests.

I will try in another environment (tried initially in Wampserver).

**Comment by dgrammatiko** [comment]
@obuisard set a break point here: https://github.com/joomla/joomla-cms/pull/39586/files#diff-b9cf1b73ab2438daa98290f2c650099e320527df061d94e31763982e9743dacdR514 do you get any errors or the sanitisation completes correctly?

**Comment by obuisard** [comment]
> @obuisard set a break point here: https://github.com/joomla/joomla-cms/pull/39586/files#diff-b9cf1b73ab2438daa98290f2c650099e320527df061d94e31763982e9743dacdR514 do you get any errors or the sanitisation completes correctly?

It does not even get there.
The only thing I can see is that, when I drag and drop a clean SVG file file under Apache with PHP 8, I get
POST /administrator/index.php?option=com_media&format=json&mediatypes=0,1,2,3&task=api.files&path=local-images:/ HTTP/1.1" 200
But under PHP 7, I get
POST /administrator/index.php?option=com_media&format=json&mediatypes=0,1,2,3&task=api.files&path=local-images:/ HTTP/1.1" 403

**Comment by dgrammatiko** [comment]
@obuisard then that's another issue

**Comment by obuisard** [comment]
> @obuisard then that's another issue

Agree, just can't figure out what it is. We need someone else to test your PR. Reaching out.

**Comment by dgrammatiko** [comment]
@obuisard can you email me you php 7.4 ini file?

**Comment by dgrammatiko** [comment]
@obuisard fwiw the issue is not related as it's (probably) coming from https://github.com/joomla/joomla-cms/blob/4dcee6991fdbbef2daa20168d930f3b4e790dcb3/administrator/components/com_media/src/Controller/ApiController.php#L188-L190 which seems a invalid token or something similar...

**Comment by sdwjoomla** [comment]
I have tested this item :red_circle: unsuccessfully on be075155dea2666c2cfb576aa4ec2e37232b7d0a<br /><br />PHP 7.4.21
Unable to upload SVG. 
Content > Media > Upload > selected image > "Upload button" 
Outcome: message "An error occurred." and image didn't upload

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39586">issues.joomla.org/tracker/joomla-cms/39586</a>.</sub>

**Comment by dgrammatiko** [comment]
@sdwjoomla can you try uploading the same svg on the 4.3-dev branch (not this pr)? 
if it’s failing there then you have another issue that this pr revealed…

**Comment by sdwjoomla** [comment]
Retesting ... actions taken and the outcome listed below ... all testing done in PHP 7.4 except where noted.

Downloaded Joomla_4.3.0-alpha3-dev+pr.39586-Development-Full_Package.zip
Image uploaded. Media manager Info does not include dimensions when on PHP7.4. but does on PHP8.0.

Clicking the vertical three dot menu and selecting Preview displays the image correctly. Thumbnail does not display.
Selecting the image as logo for the template worked and shows the correct dimensions.

Tried upload a new SVG, no change to settings, received error "unable to upload file"
Checked settings and file sizes - no issues with them.

In Cassiopeia, changed logo back to a PNG, saved, then changed back to SVG, the dimensions displayed incorrectly as 0 width and 0 height.

Manually added SVG to the images folder on the server, thumbnail displays, selected the image in the media manager and clicked the 'i' icon/link, the dimensions display correctly
Selected the new image as a logo, the dimensions display correctly as images/SimplePurpleCarTopView.svg#joomlaImage://local-images/SimplePurpleCarTopView.svg?width=148&height=73

Downloaded Joomla_4.3.0-alpha3-dev-Development-Full_Package.zip
Used https://ci.joomla.org/artifacts/joomla/joomla-cms/4.3-dev/39586/downloads/60858/pr_list.xml to apply PR
Tried to upload a SVG, no change to settings, received error "unable to upload file"
Checked settings and file sizes - no issues with them.
Manually added SVG to the images folder on the server, thumbnail displays, selected the image in the media manager and clicked the 'i' icon/link, the dimensions display correctly
Selected the new image as a logo, the dimensions display correctly as images/cyberscooty-kids_smiling.svg#joomlaImage://local-images/cyberscooty-kids_smiling.svg?width=165&height=150



Please note there is a typo in the testing instructions, image\svg+xml, should be image/svg+xml

**Comment by dgrammatiko** [comment]
@obuisard @sdwjoomla fwiw both 4.2 and 4.3 failed to upload svg with 7.4. Anyways, the problem is the returning mime type, with php 8+ we get `application/octet-stream` but with php 7 we get `image/svg` which is not valid.

Could you check, retest?

@Quy sorry can we have one more retest here?

Thanks

**Comment by obuisard** [comment]
Your last changes fixed everything on my test sites. Thumbs up for your PR, Dimitris @dgrammatiko and help in figuring out what was wrong on our tests.

**Comment by obuisard** [comment]
I have tested this item :white_check_mark: successfully on dfcf2056a6c26923b851ebf3f533ede63834c058<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39586">issues.joomla.org/tracker/joomla-cms/39586</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on dfcf2056a6c26923b851ebf3f533ede63834c058<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39586">issues.joomla.org/tracker/joomla-cms/39586</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39586">issues.joomla.org/tracker/joomla-cms/39586</a>.</sub>

**Comment by obuisard** [comment]
Thank you Dimitris @dgrammatiko for this PR!

**Comment by dgrammatiko** [comment]
@Quy @obuisard @sdwjoomla thanks for the tests
@crystalenka thanks for the preview css

@SniperSister @HLeithner please consider using a composer package for the mime types

**Pull request merged**